src/Aquifer.Common/Services/CachingLanguageService.cs

```diff
@@ -0,0 +1,75 @@
+﻿using Aquifer.Data;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.Caching.Memory;
+
+namespace Aquifer.Common.Services;
+
+public interface ICachingLanguageService
+{
+    Task<string?> GetLanguageCodeAsync(int languageId, CancellationToken ct);
+    Task<IReadOnlyDictionary<int, string>> GetLanguageCodeByIdMapAsync(CancellationToken ct);
+    Task<int?> GetLanguageIdAsync(string languageCode, CancellationToken ct);
+    Task<IReadOnlyDictionary<string, int>> GetLanguageIdByCodeMapAsync(CancellationToken ct);
+}
+
+public class CachingLanguageService(AquiferDbContext _dbContext, IMemoryCache _memoryCache) : ICachingLanguageService
+{
+    public async Task<string?> GetLanguageCodeAsync(int languageId, CancellationToken ct)
+    {
+        return (await GetLanguageCodeByIdMapAsync(ct))
+            .TryGetValue(languageId, out var languageCode)
+                ? languageCode
+                : null;
+    }
+
+    public async Task<IReadOnlyDictionary<int, string>> GetLanguageCodeByIdMapAsync(CancellationToken ct)
+    {
+        return (await GetDictionariesFromCacheAsync(ct))
+            .LanguageCodeByIdMap;
+    }
+
+    public async Task<int?> GetLanguageIdAsync(string languageCode, CancellationToken ct)
+    {
+        ArgumentNullException.ThrowIfNull(languageCode);
+
+        return (await GetLanguageIdByCodeMapAsync(ct))
+            .TryGetValue(languageCode, out var languageId)
+                ? languageId
+                : null;
```

Comment:
```suggestion
return (await GetLanguageCodeByIdMapAsync(ct)).GetValueOrDefault(languageId);
```

Reply:
The default value of the cached dictionary is `default<int>` which is `0`.  Thus I used the `TryGetValue` strategy so that I could return `null` instead for the default to represent the fact that the there is no available ID for the given language code (and vice versa for the other method above).

You would _think_ that we could still use `GetValueOrDefault()` with a `defaultValue` of `null` in that scenario like so:
```
return (await GetLanguageCodeByIdMapAsync(ct)).GetValueOrDefault(languageId, defaultValue: null);
```
but it unfortunately doesn't work due to the return type of that extension method not being nullable.

Reply:
Ah, that explains why Rider only suggested the other one be updated, and not this one (I accidentally put the suggestion on this one, when I meant to put it on the one above). I was wondering what the difference was. So the other one can use `GetValueOrDefault` since string is nullable, but this one can't.

Reply:
The Dictionary would have to have `string?` but it uses `string` so the other usage also can't use `GetValueOrDefault()`.